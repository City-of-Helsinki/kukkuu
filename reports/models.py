from django.db import models
from django.utils.translation import gettext_lazy as _

ACCESS_REPORT_API_PERM = "reports.access_report_api"


class Permission(models.Model):
    """Non-db model just for holding reports' permissions.

    Needed because all Django's permissions need to be tied to a Model/ContentType.
    """

    class Meta:
        managed = False
        default_permissions = []
        permissions = [
            (ACCESS_REPORT_API_PERM.split(".")[1], _("Can access report API"))
        ]

    @classmethod
    def get_codenames(cls):
        return [codename for codename, _ in cls._meta.permissions]
