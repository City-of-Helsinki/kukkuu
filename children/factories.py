import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
