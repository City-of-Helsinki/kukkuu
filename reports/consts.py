OTHER_LANGUAGE_API_NAME = "__OTHER__"


# Ideally we should get these from the database, as it is theoretically possible we
# might want add more languages dynamically in the future, but there doesn't seem to be
# an easy way to do that in a way that works with DRF Spectacular.
LANGUAGE_CHOICES = [
    ("ara", "Arabic"),
    ("ben", "Bengali"),
    ("deu", "German"),
    ("eng", "English"),
    ("est", "Estonian"),
    ("fas", "Persian"),
    ("fin", "Finnish"),
    ("fra", "French"),
    ("hin", "Hindi"),
    ("ita", "Italian"),
    ("kur", "Kurdish"),
    ("nep", "Nepali"),
    ("nor", "Norwegian"),
    ("pol", "Polish"),
    ("por", "Portuguese"),
    ("ron", "Romanian"),
    ("rus", "Russian"),
    ("smi", "Sami"),
    ("som", "Somali"),
    ("spa", "Spanish"),
    ("sqi", "Albanian"),
    ("swe", "Swedish"),
    ("tgl", "Tagalog"),
    ("tha", "Thai"),
    ("tur", "Turkish"),
    ("urd", "Urdu"),
    ("vie", "Vietnamese"),
    ("zho", "Chinese"),
    (OTHER_LANGUAGE_API_NAME, "Other language"),
]

CONTACT_LANGUAGE_TO_LANGUAGE = {
    "fi": "fin",
    "sv": "swe",
    "en": "eng",
}

CONTACT_LANGUAGE_PRIORITIES = ["fi", "en", "sv"]
