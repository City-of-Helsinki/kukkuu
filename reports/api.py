from datetime import date
from typing import List, Optional

from django.db.models import Prefetch
from django.utils.timezone import localdate
from drf_spectacular.utils import (
    extend_schema,
    extend_schema_field,
    extend_schema_serializer,
    extend_schema_view,
    OpenApiExample,
)
from rest_framework import mixins, serializers, viewsets

from children.models import Child
from languages.models import Language

OTHER_LANGUAGE_API_NAME = "__OTHER__"


# Ideally we should get these from the database, as it is theoretically possible we
# might want add more languages dynamically in the future, but there doesn't seem to be
# an easy way to do that in a way that works with DRF Spectacular.
LANGUAGE_CHOICES = [
    ("ara", "Arabic"),
    ("ben", "Bengali"),
    ("deu", "German"),
    ("eng", "English"),
    ("est", "Estonian"),
    ("fas", "Persian"),
    ("fin", "Finnish"),
    ("fra", "French"),
    ("hin", "Hindi"),
    ("ita", "Italian"),
    ("kur", "Kurdish"),
    ("nep", "Nepali"),
    ("nor", "Norwegian"),
    ("pol", "Polish"),
    ("por", "Portuguese"),
    ("ron", "Romanian"),
    ("rus", "Russian"),
    ("smi", "Sami"),
    ("som", "Somali"),
    ("spa", "Spanish"),
    ("sqi", "Albanian"),
    ("swe", "Swedish"),
    ("tgl", "Tagalog"),
    ("tha", "Thai"),
    ("tur", "Turkish"),
    ("urd", "Urdu"),
    ("vie", "Vietnamese"),
    ("zho", "Chinese"),
    (OTHER_LANGUAGE_API_NAME, "Other language"),
]

CONTACT_LANGUAGE_TO_LANGUAGE = {
    "fi": "fin",
    "sv": "swe",
    "en": "eng",
}

CONTACT_LANGUAGE_PRIORITIES = ["fi", "en", "sv"]


def get_primary_contact_language(contact_languages: List[str]):
    return [
        contact_lang
        for lang in CONTACT_LANGUAGE_PRIORITIES
        for contact_lang in contact_languages
        if contact_lang == lang
    ][0]


@extend_schema_serializer(
    examples=[
        OpenApiExample(
            "Example 1",
            value=[
                {
                    "registration_date": "2021-02-18",
                    "birth_year": 2021,
                    "contact_language": "fin",
                    "languages_spoken_at_home": ["fin", "nor", "__OTHER__"],
                },
                {
                    "registration_date": "2021-03-18",
                    "birth_year": 2020,
                    "contact_language": "eng",
                    "languages_spoken_at_home": [],
                },
            ],
            response_only=True,
        ),
    ]
)
class ChildSerializer(serializers.ModelSerializer):
    child_birthdate_postal_code_guardian_emails_hash = serializers.SerializerMethodField(  # noqa
        help_text="Salted hash of child's birthdate, postal code and guardians' emails."
    )
    child_name_birthdate_postal_code_guardian_emails_hash = (
        serializers.SerializerMethodField(
            help_text="Salted hash of child's name (i.e. first name and last name), "
            "birthdate, postal code and guardians' emails."
        )
    )
    registration_date = serializers.SerializerMethodField()
    birth_year = serializers.SerializerMethodField()
    contact_language = serializers.SerializerMethodField()
    languages_spoken_at_home = serializers.SerializerMethodField(
        help_text="Array of ISO 639-3 (language) or ISO 639-5 (language family) "
        "alpha-3 codes. Value `__OTHER__` means any other language."  # noqa
    )

    class Meta:
        model = Child
        fields = [
            "child_birthdate_postal_code_guardian_emails_hash",
            "child_name_birthdate_postal_code_guardian_emails_hash",
            "registration_date",
            "birth_year",
            "postal_code",
            "contact_language",
            "languages_spoken_at_home",
        ]

    def get_child_birthdate_postal_code_guardian_emails_hash(self, child: Child) -> str:
        """
        Salted hash of child's birthdate, postal code and guardians' emails.
        """
        return child.birthdate_postal_code_guardian_emails_hash

    def get_child_name_birthdate_postal_code_guardian_emails_hash(
        self, child: Child
    ) -> str:
        """
        Salted hash of child's name (i.e. first name and last name), birthdate, postal
        code and guardians' emails.
        """
        return child.name_birthdate_postal_code_guardian_emails_hash

    def get_registration_date(self, obj: Child) -> date:
        return localdate(obj.created_at)

    def get_birth_year(self, obj: Child) -> int:
        return obj.birthdate.year

    @extend_schema_field(
        serializers.ChoiceField(
            choices=(("fin", "Finnish"), ("swe", "Swedish"), ("eng", "English"))
        )
    )
    def get_contact_language(self, obj: Child) -> str:
        contact_languages = [
            guardian.language for guardian in list(obj.guardians.all())
        ]
        primary_contact_lang = get_primary_contact_language(contact_languages)
        return CONTACT_LANGUAGE_TO_LANGUAGE[primary_contact_lang]

    @extend_schema_field(
        serializers.ListField(child=serializers.ChoiceField(choices=LANGUAGE_CHOICES))
    )
    def get_languages_spoken_at_home(self, obj: Child) -> List[Optional[str]]:
        return [
            l.alpha_3_code or OTHER_LANGUAGE_API_NAME
            for l in obj.guardians.all()[0].prefetched_languages_spoken_at_home
        ]


@extend_schema_view(list=extend_schema(description="Get all children data."))
class ChildViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
    queryset = (
        Child.objects.exclude(guardians=None)
        .prefetch_related(
            Prefetch(
                "guardians__languages_spoken_at_home",
                queryset=Language.objects.order_by("alpha_3_code"),
                to_attr="prefetched_languages_spoken_at_home",
            ),
        )
        .prefetch_related("guardians")
        .order_by("name")
    )
    serializer_class = ChildSerializer
