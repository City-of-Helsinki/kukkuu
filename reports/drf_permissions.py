from rest_framework.permissions import BasePermission

from reports.models import ACCESS_REPORT_API_PERM


class ReportAPIPermission(BasePermission):
    def has_permission(self, request, view):
        return request.user.has_perm(ACCESS_REPORT_API_PERM)
