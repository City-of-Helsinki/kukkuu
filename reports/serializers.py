from datetime import date
from typing import Any, List, Optional

from django.utils.timezone import localdate
from drf_spectacular.utils import (
    extend_schema_field,
)
from rest_framework import serializers

from children.models import Child
from events.models import Event, EventGroup, Occurrence
from projects.models import Project
from reports.consts import (
    CONTACT_LANGUAGE_PRIORITIES,
    CONTACT_LANGUAGE_TO_LANGUAGE,
    LANGUAGE_CHOICES,
    OTHER_LANGUAGE_API_NAME,
)
from venues.models import Venue


def get_primary_contact_language(contact_languages: List[str]):
    return [
        contact_lang
        for lang in CONTACT_LANGUAGE_PRIORITIES
        for contact_lang in contact_languages
        if contact_lang == lang
    ][0]


class DynamicFieldsModelSerializer(serializers.ModelSerializer):
    """
    A ModelSerializer that takes an additional `fields` argument that
    controls which fields should be displayed.
    """

    def __init__(self, *args, **kwargs):
        # Don't pass the 'fields' arg up to the superclass
        fields = kwargs.pop("fields", None)

        # Instantiate the superclass normally
        super().__init__(*args, **kwargs)

        if fields is not None:
            # Drop any fields that are not specified in the `fields` argument.
            allowed = set(fields)
            existing = set(self.fields)
            for field_name in existing - allowed:
                self.fields.pop(field_name)


class ChildSerializer(serializers.ModelSerializer):
    child_birthyear_postal_code_guardian_emails_hash = serializers.SerializerMethodField(  # noqa
        help_text="Salted hash of child's birthyear, postal code and guardians' emails."
    )
    child_name_birthyear_postal_code_guardian_emails_hash = (
        serializers.SerializerMethodField(
            help_text="Salted hash of child's name (i.e. first name and last name), "
            "birthyear, postal code and guardians' emails."
        )
    )
    registration_date = serializers.SerializerMethodField()
    birthyear = serializers.SerializerMethodField()
    contact_language = serializers.SerializerMethodField()
    languages_spoken_at_home = serializers.SerializerMethodField(
        help_text="Array of ISO 639-3 (language) or ISO 639-5 (language family) "
        "alpha-3 codes. Value `__OTHER__` means any other language."  # noqa
    )
    is_obsolete = serializers.SerializerMethodField(
        help_text="Are all the child's guardian user instances marked as obsolete?"
    )

    class Meta:
        model = Child
        fields = [
            "child_birthyear_postal_code_guardian_emails_hash",
            "child_name_birthyear_postal_code_guardian_emails_hash",
            "registration_date",
            "birthyear",
            "postal_code",
            "contact_language",
            "languages_spoken_at_home",
            "is_obsolete",
        ]

    def get_child_birthyear_postal_code_guardian_emails_hash(self, child: Child) -> str:
        """
        Salted hash of child's birthyear, postal code and guardians' emails.
        """
        return child.birthyear_postal_code_guardian_emails_hash

    def get_child_name_birthyear_postal_code_guardian_emails_hash(
        self, child: Child
    ) -> str:
        """
        Salted hash of child's name (i.e. first name and last name), birthyear, postal
        code and guardians' emails.
        """
        return child.name_birthyear_postal_code_guardian_emails_hash

    def get_registration_date(self, obj: Child) -> date:
        return localdate(obj.created_at)

    def get_birthyear(self, obj: Child) -> int:
        return obj.birthyear

    @extend_schema_field(
        serializers.ChoiceField(
            choices=(("fin", "Finnish"), ("swe", "Swedish"), ("eng", "English"))
        )
    )
    def get_contact_language(self, obj: Child) -> str:
        contact_languages = [
            guardian.language for guardian in list(obj.guardians.all())
        ]
        primary_contact_lang = get_primary_contact_language(contact_languages)
        return CONTACT_LANGUAGE_TO_LANGUAGE[primary_contact_lang]

    @extend_schema_field(
        serializers.ListField(child=serializers.ChoiceField(choices=LANGUAGE_CHOICES))
    )
    def get_languages_spoken_at_home(self, obj: Child) -> List[Optional[str]]:
        return [
            lang.alpha_3_code or OTHER_LANGUAGE_API_NAME
            for lang in obj.guardians.all()[0].prefetched_languages_spoken_at_home
        ]

    def get_is_obsolete(self, obj: Child) -> bool:
        return obj.is_obsolete


class ProjectSerializer(DynamicFieldsModelSerializer):
    name = serializers.SerializerMethodField()

    class Meta:
        model = Project
        fields = ["id", "year", "name"]

    # NOTE: Fetching the translations from django-parler table
    # can cause lots of sub queries.
    def get_name(self, obj: Project):
        return obj.name_with_translations


class VenueSerializer(
    DynamicFieldsModelSerializer, serializers.HyperlinkedModelSerializer
):
    name = serializers.SerializerMethodField()
    address = serializers.SerializerMethodField()
    project = ProjectSerializer(
        # To optimize the query, exclude django-parler translated fields and
        # offer url for details instead
        fields=[
            "id",
            "year",
        ],
        read_only=True,
        help_text="A project describes the year group where "
        "the children can participate. A venue is a project specific object.",
    )

    class Meta:
        model = Venue
        fields = ["url", "id", "project", "name", "address"]

    # NOTE: Fetching the translations from django-parler table
    # can cause lots of sub queries.
    def get_name(self, obj: Venue):
        return obj.name_with_translations

    # NOTE: Fetching the translations from django-parler table
    # can cause lots of sub queries.
    def get_address(self, obj: Venue):
        return obj.address_with_translations


class OccurrenceSerializer(serializers.ModelSerializer):
    enrolment_count = serializers.SerializerMethodField(
        help_text="Count of how many children have enrolled "
        "in this occurrence of an event."
    )

    attended_count = serializers.SerializerMethodField(
        help_text="Count of how many children atteded in this occurrence of an event."
    )

    capacity = serializers.SerializerMethodField(
        help_text="The total capacity of how many children can there participate "
        "in this occurrence of an event."
    )

    venue = VenueSerializer(
        read_only=True,
        # To optimize the query, exclude django-parler translated fields and
        # offer url for details instead
        fields=[
            "url",
            "id",
        ],
    )

    class Meta:
        model = Occurrence
        fields = [
            "id",
            "time",
            "venue",
            "enrolment_count",
            "attended_count",
            "capacity",
        ]

    def get_enrolment_count(self, obj: Occurrence) -> int:
        return obj.get_enrolment_count()

    def get_attended_count(self, obj: Occurrence) -> int:
        return obj.get_attended_enrolment_count()

    def get_capacity(self, obj: Occurrence) -> Optional[int]:
        return obj.get_capacity()


class EventGroupSerializer(
    DynamicFieldsModelSerializer, serializers.HyperlinkedModelSerializer
):
    name = serializers.SerializerMethodField()
    project = ProjectSerializer(
        # To optimize the query, exclude django-parler translated fields and
        # offer url for details instead
        fields=[
            "id",
            "year",
        ],
        read_only=True,
        help_text="A project describes the year group where "
        "the children can participate.",
    )
    events_count = serializers.SerializerMethodField(
        help_text="Total count of events in this event group."
    )
    enrolment_count = serializers.SerializerMethodField(
        help_text="Total count of how many children have enrolled "
        "in this event group's event occurrences."
    )
    attended_count = serializers.SerializerMethodField(
        help_text="Count of how many children atteded "
        "in this event group's event occurrences."
    )
    capacity = serializers.SerializerMethodField(
        help_text="The total capacity of how many children can there participate "
        "in this event group's event occurrences."
    )

    class Meta:
        model = EventGroup
        fields = [
            "url",
            "id",
            "name",
            "project",
            "events_count",
            "enrolment_count",
            "attended_count",
            "capacity",
        ]

    def get_events_count(self, obj: EventGroup) -> int:
        return obj.events.count()

    def get_enrolment_count(self, obj: EventGroup) -> int:
        return obj.get_enrolment_count()

    def get_attended_count(self, obj: EventGroup) -> int:
        return obj.get_attended_enrolment_count()

    def get_capacity(self, obj: EventGroup) -> Optional[int]:
        return obj.get_capacity()

    # NOTE: Fetching the translations from django-parler table
    # can cause lots of sub queries.
    def get_name(self, obj: Project):
        return obj.name_with_translations


class EventSerializer(serializers.ModelSerializer):
    name = serializers.SerializerMethodField()
    project = ProjectSerializer(
        # To optimize the query, exclude django-parler translated fields and
        # offer url for details instead
        fields=[
            "id",
            "year",
        ],
        read_only=True,
        help_text="A project describes the year group where "
        "the children can participate.",
    )
    event_group = EventGroupSerializer(
        # To optimize the query, exclude django-parler translated fields and
        # offer url for details instead
        fields=[
            "url",
            "id",
        ],
        read_only=True,
        help_text="The event can belong to an event group "
        "that contains a group of events.",
    )
    occurrences = OccurrenceSerializer(
        many=True,
        read_only=True,
        help_text="A single event can contain 1 or multiple occurrences "
        "that are held at different times",
    )
    occurrences_count = serializers.SerializerMethodField(
        help_text="Total count of occurrences in this event."
    )
    enrolment_count = serializers.SerializerMethodField(
        help_text="Total count of how many children have enrolled "
        "in this event's occurrences."
    )
    attended_count = serializers.SerializerMethodField(
        help_text="Count of how many children atteded in this event's occurrences."
    )
    capacity = serializers.SerializerMethodField(
        help_text="The total capacity of how many children can there participate "
        "in this event's occurrences."
    )

    class Meta:
        model = Event
        fields = [
            "id",
            "name",
            "project",
            "event_group",
            "occurrences",
            "ticket_system",
            "occurrences_count",
            "enrolment_count",
            "attended_count",
            "capacity",
        ]

    def get_occurrences_count(self, obj: Event) -> int:
        return obj.occurrences.count()

    def get_enrolment_count(self, obj: Event) -> int:
        return obj.get_enrolment_count()

    def get_attended_count(self, obj: Event) -> int:
        return obj.get_attended_enrolment_count()

    def get_capacity(self, obj: Event) -> Optional[int]:
        return obj.get_capacity()

    # NOTE: Fetching the translations from django-parler table
    # can cause lots of sub queries.
    def get_name(self, obj: Project) -> dict[Any, str]:
        return obj.name_with_translations
