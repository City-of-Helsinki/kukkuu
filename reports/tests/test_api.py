import base64
from datetime import date

import freezegun
import pytest
from django.contrib.auth import get_user_model
from django.test import override_settings
from guardian.shortcuts import assign_perm, remove_perm
from rest_framework.test import APIClient

from children.factories import (
    ChildFactory,
    ChildWithGuardianFactory,
    ChildWithTwoGuardiansFactory,
)
from languages.models import Language
from reports.api import get_primary_contact_language
from reports.drf_permissions import ACCESS_REPORT_API_PERM

LIST_ENDPOINT = "/reports/children/"


@pytest.fixture(autouse=True)
def languages():
    for code in ("fin", "swe", "eng"):
        Language.objects.create_from_language_code(code)
    Language.objects.create_option_other()


@pytest.fixture
def user():
    user = get_user_model().objects.create_user(
        username="test_user",
        password="test_password",
    )
    assign_perm(ACCESS_REPORT_API_PERM, user)
    return user


@pytest.fixture
def user_api_client(user):
    api_client = APIClient()
    api_client.user = user
    credentials = b"test_user:test_password"
    api_client.credentials(
        HTTP_AUTHORIZATION=f"Basic {base64.b64encode(credentials).decode('ascii')}"
    )
    return api_client


@pytest.mark.django_db
def test_children_endpoint_not_authenticated(user):
    response = APIClient().get(LIST_ENDPOINT)
    assert response.status_code == 401


@pytest.mark.django_db
def test_children_endpoint_no_permission(user_api_client):
    remove_perm(ACCESS_REPORT_API_PERM, user_api_client.user)
    response = user_api_client.get(LIST_ENDPOINT)
    assert response.status_code == 403


@pytest.mark.django_db
@override_settings(
    # Must be the same as in CI pipeline for the snapshot to be an exact match:
    KUKKUU_HASHID_SALT="almosttopsecret123"
)
def test_children_endpoint(user_api_client, snapshot, django_assert_max_num_queries):
    with freezegun.freeze_time("2021-02-02T12:00:00Z"):
        ChildWithGuardianFactory(
            first_name="Doe John",
            birthdate=date(2021, 1, 1),
            postal_code="11111",
            relationship__guardian__language="fi",
            relationship__guardian__languages_spoken_at_home=["fin", None],
            relationship__guardian__email="advocate@example.org",
        )

    with freezegun.freeze_time("2021-03-03T01:00:00Z"):
        ChildWithGuardianFactory(
            first_name="Garcia Isabella",
            birthdate=date(2021, 2, 2),
            postal_code="22222",
            relationship__guardian__language="sv",
            relationship__guardian__languages_spoken_at_home=[None],
            relationship__guardian__email="parent@example.org",
        )

    # This guy's registration date should be 2021-04-05 in Helsinki time
    with freezegun.freeze_time("2021-04-04T23:00:00Z"):
        ChildWithGuardianFactory(
            first_name="Doe Jane",
            birthdate=date(2020, 3, 3),
            postal_code="33333",
            relationship__guardian__language="en",
            relationship__guardian__email="something@to-fool-factoryboy.net",
        )

    with freezegun.freeze_time("2021-04-05T21:05:00Z"):
        ChildWithTwoGuardiansFactory(
            first_name="Last name First name",
            birthdate=date(2020, 3, 3),
            postal_code="44444",
            relationship__guardian__language="en",
            relationship__guardian__email="guardian@example.org",
            relationship2__guardian__language="sv",
            relationship2__guardian__email="advocate@example.org",
        )

    ChildFactory()  # This is an orphan child so she should not be in the results

    with django_assert_max_num_queries(6):
        response = user_api_client.get(LIST_ENDPOINT)
    assert response.status_code == 200, response.content

    json_data = response.json()

    # The result set should be ordered by 'first_name'
    assert [child["postal_code"] for child in json_data] == [
        "33333",
        "11111",
        "22222",
        "44444",
    ]

    snapshot.assert_match(json_data)


@pytest.mark.parametrize(
    "contact_languages,language",
    [
        [["en", "sv", "fi"], "fi"],
        [["fi", "en", "sv"], "fi"],
        [["en", "sv"], "en"],
        [["en"], "en"],
        [["fi", "en"], "fi"],
        [["sv"], "sv"],
    ],
)
@pytest.mark.django_db
def test_get_primary_contact_language(contact_languages, language):
    assert get_primary_contact_language(contact_languages) == language
