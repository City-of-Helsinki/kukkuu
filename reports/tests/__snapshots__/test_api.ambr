# serializer version: 1
# name: test_children_endpoint
  list([
    dict({
      'birthyear': 2020,
      'child_birthyear_postal_code_guardian_emails_hash': '3b78fc2b60381829414bd2bd9ea53e90123d9cd7faeb35f86f5038d669dd96e0',
      'child_name_birthyear_postal_code_guardian_emails_hash': 'bd94f9a87d842f6889f10aa72528a273c27fb864fe7b14b7b38b2b37a59b615a',
      'contact_language': 'eng',
      'is_obsolete': False,
      'languages_spoken_at_home': list([
      ]),
      'postal_code': '33333',
      'registration_date': '2021-04-05',
    }),
    dict({
      'birthyear': 2021,
      'child_birthyear_postal_code_guardian_emails_hash': 'd9653e3210babd7c8d6094f069ddfbf790d6b16d368d0b14e0d4f9d1d2eed8aa',
      'child_name_birthyear_postal_code_guardian_emails_hash': 'cde8b66cfb17f9dd1e3a27b34ced099567555c09ac53ab8d96eb6f108b481a78',
      'contact_language': 'fin',
      'is_obsolete': False,
      'languages_spoken_at_home': list([
        'fin',
        '__OTHER__',
      ]),
      'postal_code': '11111',
      'registration_date': '2021-02-02',
    }),
    dict({
      'birthyear': 2021,
      'child_birthyear_postal_code_guardian_emails_hash': '73471fad7bbcffa243a695f211ffd2c0ddb89531d747548709c3677ecbe4a0cc',
      'child_name_birthyear_postal_code_guardian_emails_hash': 'c056d2e1b6ca2ab0189ce068567cb6d3951928283e8ebbb384ec20d3a85f632e',
      'contact_language': 'swe',
      'is_obsolete': False,
      'languages_spoken_at_home': list([
        '__OTHER__',
      ]),
      'postal_code': '22222',
      'registration_date': '2021-03-03',
    }),
    dict({
      'birthyear': 2020,
      'child_birthyear_postal_code_guardian_emails_hash': 'f1694b9c7d7a45c8606d9506f171025079940fcc97d5e125e1cc73495461f378',
      'child_name_birthyear_postal_code_guardian_emails_hash': '92d10305295a1210620ae3158c5aeecc24d9018be279042e8226d120cbfdb978',
      'contact_language': 'eng',
      'is_obsolete': False,
      'languages_spoken_at_home': list([
      ]),
      'postal_code': '44444',
      'registration_date': '2021-04-06',
    }),
  ])
# ---
# name: test_event_groups_endpoint
  list([
    dict({
      'attended_count': 0,
      'capacity': 20,
      'enrolment_count': 1,
      'events_count': 1,
      'id': 1,
      'name': dict({
        'en': '',
        'fi': 'Example Event Group 1',
        'sv': '',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/event-group/1/',
    }),
    dict({
      'attended_count': 0,
      'capacity': 0,
      'enrolment_count': 0,
      'events_count': 0,
      'id': 2,
      'name': dict({
        'en': 'Example Event Group 2 (en)',
        'fi': 'Example Event Group 2 (fi)',
        'sv': 'Example Event Group 2 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/event-group/2/',
    }),
    dict({
      'attended_count': 0,
      'capacity': 0,
      'enrolment_count': 0,
      'events_count': 0,
      'id': 3,
      'name': dict({
        'en': 'Example Event Group 3 (en)',
        'fi': 'Example Event Group 3 (fi)',
        'sv': 'Example Event Group 3 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/event-group/3/',
    }),
    dict({
      'attended_count': 0,
      'capacity': 0,
      'enrolment_count': 0,
      'events_count': 0,
      'id': 4,
      'name': dict({
        'en': 'Example Event Group 4 (en)',
        'fi': 'Example Event Group 4 (fi)',
        'sv': 'Example Event Group 4 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/event-group/4/',
    }),
    dict({
      'attended_count': 0,
      'capacity': 0,
      'enrolment_count': 0,
      'events_count': 0,
      'id': 5,
      'name': dict({
        'en': 'Example Event Group 5 (en)',
        'fi': 'Example Event Group 5 (fi)',
        'sv': 'Example Event Group 5 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/event-group/5/',
    }),
  ])
# ---
# name: test_event_groups_retrieve_endpoint
  dict({
    'attended_count': 1,
    'capacity': 20,
    'enrolment_count': 2,
    'events_count': 1,
    'id': 1,
    'name': dict({
      'en': 'Example Event Group (en)',
      'fi': 'Example Event Group (fi)',
      'sv': 'Example Event Group (sv)',
    }),
    'project': dict({
      'id': 1,
      'year': 2020,
    }),
    'url': 'http://testserver/reports/event-group/1/',
  })
# ---
# name: test_events_endpoint
  list([
    dict({
      'attended_count': 0,
      'capacity': 40,
      'enrolment_count': 2,
      'event_group': dict({
        'id': 1,
        'url': 'http://testserver/reports/event-group/1/',
      }),
      'id': 1,
      'name': dict({
        'en': '',
        'fi': 'Write century spring never skill down subject town.',
        'sv': '',
      }),
      'occurrences': list([
        dict({
          'attended_count': 0,
          'capacity': 20,
          'enrolment_count': 2,
          'id': 1,
          'time': '2021-04-05T02:00:00+03:00',
          'venue': dict({
            'id': 1,
            'url': 'http://testserver/reports/venue/1/',
          }),
        }),
        dict({
          'attended_count': 0,
          'capacity': 20,
          'enrolment_count': 0,
          'id': 2,
          'time': '2021-04-05T02:00:00+03:00',
          'venue': dict({
            'id': 2,
            'url': 'http://testserver/reports/venue/2/',
          }),
        }),
      ]),
      'occurrences_count': 2,
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'ticket_system': 'internal',
    }),
    dict({
      'attended_count': 0,
      'capacity': 80,
      'enrolment_count': 2,
      'event_group': None,
      'id': 2,
      'name': dict({
        'en': '',
        'fi': 'People sense knowledge stock subject.',
        'sv': '',
      }),
      'occurrences': list([
        dict({
          'attended_count': 0,
          'capacity': 40,
          'enrolment_count': 2,
          'id': 3,
          'time': '2021-04-05T02:00:00+03:00',
          'venue': dict({
            'id': 3,
            'url': 'http://testserver/reports/venue/3/',
          }),
        }),
        dict({
          'attended_count': 0,
          'capacity': 40,
          'enrolment_count': 0,
          'id': 4,
          'time': '2021-04-05T02:00:00+03:00',
          'venue': dict({
            'id': 4,
            'url': 'http://testserver/reports/venue/4/',
          }),
        }),
      ]),
      'occurrences_count': 2,
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'ticket_system': 'internal',
    }),
  ])
# ---
# name: test_venue_retrieve_endpoint
  dict({
    'address': dict({
      'en': 'Example Address (en)',
      'fi': 'Example Address (fi)',
      'sv': 'Example Address (sv)',
    }),
    'id': 1,
    'name': dict({
      'en': 'Example Venue (en)',
      'fi': 'Example Venue (fi)',
      'sv': 'Example Venue (sv)',
    }),
    'project': dict({
      'id': 1,
      'year': 2020,
    }),
    'url': 'http://testserver/reports/venue/1/',
  })
# ---
# name: test_venues_endpoint
  list([
    dict({
      'address': dict({
        'en': 'Example Address 1 (en)',
        'fi': 'Example Address 1 (fi)',
        'sv': 'Example Address 1 (sv)',
      }),
      'id': 1,
      'name': dict({
        'en': 'Example Venue 1 (en)',
        'fi': 'Example Venue 1 (fi)',
        'sv': 'Example Venue 1 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/venue/1/',
    }),
    dict({
      'address': dict({
        'en': 'Example Address 2 (en)',
        'fi': 'Example Address 2 (fi)',
        'sv': 'Example Address 2 (sv)',
      }),
      'id': 2,
      'name': dict({
        'en': 'Example Venue 2 (en)',
        'fi': 'Example Venue 2 (fi)',
        'sv': 'Example Venue 2 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/venue/2/',
    }),
    dict({
      'address': dict({
        'en': 'Example Address 3 (en)',
        'fi': 'Example Address 3 (fi)',
        'sv': 'Example Address 3 (sv)',
      }),
      'id': 3,
      'name': dict({
        'en': 'Example Venue 3 (en)',
        'fi': 'Example Venue 3 (fi)',
        'sv': 'Example Venue 3 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/venue/3/',
    }),
    dict({
      'address': dict({
        'en': 'Example Address 4 (en)',
        'fi': 'Example Address 4 (fi)',
        'sv': 'Example Address 4 (sv)',
      }),
      'id': 4,
      'name': dict({
        'en': 'Example Venue 4 (en)',
        'fi': 'Example Venue 4 (fi)',
        'sv': 'Example Venue 4 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/venue/4/',
    }),
    dict({
      'address': dict({
        'en': 'Example Address 5 (en)',
        'fi': 'Example Address 5 (fi)',
        'sv': 'Example Address 5 (sv)',
      }),
      'id': 5,
      'name': dict({
        'en': 'Example Venue 5 (en)',
        'fi': 'Example Venue 5 (fi)',
        'sv': 'Example Venue 5 (sv)',
      }),
      'project': dict({
        'id': 1,
        'year': 2020,
      }),
      'url': 'http://testserver/reports/venue/5/',
    }),
  ])
# ---
